include/nbl/asset/IDescriptorSet.h

```diff
@@ -47,9 +47,10 @@ class IDescriptorSet : public virtual core::IReferenceCounted // TODO: try to re
                 };
                 struct SImageInfo
                 {
-					// This will be ignored if the DS layout already has an immutable sampler specified for the binding.
-                    core::smart_refctd_ptr<typename layout_t::sampler_type> sampler;
-                    IImage::LAYOUT imageLayout;
+					IImage::LAYOUT imageLayout;
+					// If the binding is COMBINED, this will be ignored if the DS layout already has an immutable sampler specified for the binding
+					// If the binding is an immutable SAMPLER, whatever's here doesn't matter because the write validation will fail
+					core::smart_refctd_ptr<typename layout_t::sampler_type> sampler;
```

Comment:
btw have you established if its invalid to have `sampler` non-nullptr and `desc` non-nullptr when `desc` is a SAMPLER ?

Reply:
No, and in fact I just realized both `sampler` and `desc` are either pointing to the same thing pointlessly, or could point to different things. 

The latter is unwanted since it's flatout wrong so I'll refactor so it doesn't happen. I think I'll opt for forcing `sampler` to be nullptr when the descriptor is a SAMPLER, since most predicated for behaviour are done by checking the category of `desc`

Reply:
Forcing `sampler` and `desc` to point to the same object in such a case is also an option - I get an extra useless smart pointer but the code ends up being simpler I think: no edge cases for the standalone sampler type

Reply:
Okay I think I can have both: I give `SDescriptorInfo` some sort of `getSampler()` method that I can use throughout the rest of the code so I don't have edge cases, and also force SAMPLER to have its `info.sampler` be nullptr or at least just ignore that field

Reply:
So now, a question about write/copy validation: I expect to only be using `desc` for descriptors of type SAMPLER, and ignore whatever is given under the `info` field. Should I, however, do anything about the info field? Like do I force it to be nullptr and drop a smart pointer there, or is that a problem of the user?

Reply:
yeah it seems like `info.sampler` should be expected to be `nullptr` when the `desc` is `SAMPLER`

and nobody should ever be providing a non-null `info.sampler` for `COMBINED` binding with immutable sampler, or any sbindinginfo at all for a sampler binding that has an immutable sampler

You can also rename the `samplers` to `immutableSamplers` here for clarity https://github.com/Devsh-Graphics-Programming/Nabla/blob/3df38a45978510f9823c97adfbc3bea4c90e8aee/include/nbl/asset/IDescriptorSetLayout.h#L62

Reply:
> Okay I think I can have both: I give `SDescriptorInfo` some sort of `getSampler()` method that I can use throughout the rest of the code so I don't have edge cases, and also force SAMPLER to have its `info.sampler` be nullptr or at least just ignore that field

> So now, a question about write/copy validation: I expect to only be using desc for descriptors of type SAMPLER, and ignore whatever is given under the info field. Should I, however, do anything about the info field? Like do I force it to be nullptr and drop a smart pointer there, or is that a problem of the user?

its a union with a buffer, its just sanest to ignore them like we do for buffers.

For semantical clarity you can define everything as:
```cpp
struct SDescriptorInfo
{
  struct SBufferInfo
  {
    size_t offset;
    size_t size;//in Vulkan it's called `range` but IMO it's misleading so i changed to `size`
    
    auto operator<=>(const SBufferInfo&) const = default;
  };
  struct SImageInfo
  {
    IImage::LAYOUT imageLayout;
  };
  struct SCombinedImageSamplerInfo : SImageInfo
  {
    core::smart_refctd_ptr<typename layout_t::sampler_type> sampler;
  };

  core::smart_refctd_ptr<IDescriptor> desc;
  union SBufferImageInfo
  {
    SBufferImageInfo()
    {
      memset(&buffer, 0, core::max<size_t>(sizeof(buffer), sizeof(combinedImageSampler)));
    };

    ~SBufferImageInfo() {};

    SBufferInfo buffer;
    SImageInfo image;
    SCombinedImageSamplerInfo combinedImageSampler;
  } info;
};
```

Reply:
This would break a lot of existing code potentially, is it okay to do this?

Reply:
> This would break a lot of existing code potentially, is it okay to do this?

how would it break? its still a union and bytewise the members are still in the same place.

only `image` turns into `combinedImageSampler` but thats an easy `Ctrl+H` job